### ISSUE TITLE
Problem using the glmnet logistic regression model trained in Caret but with only the features selected predictor variables

### ISSUE THREAD

**Issue by samleighton87** [topepo/caret#978]
I have trained a glmnet binary logistic regression model in caret by tuning over a grid of lambda at alpha 0.5 and selecting for the sparsest model within oneSE of the best trained performance. It has selected features. Predict.train in caret automatically applies my preprocessing (as specified during training) to my new test data. However, it requires the same columns as the train data. 

I would like to be able to input only the feature selected predictor variables into the model and get a prediction. 

I tried to do this manually by extracting the final model coefficients at the optimally tuned lambda, manually standardising the predictor variables and calculating the logit then probability for my outcome. 

However, I get a different probability for my outcome compared to using predict.train function. I am not really sure where I am going wrong.

I will outline my problem with the following example using the Titanic dataset:

    #load required libraries
    library(plyr)
    library(caret)
    library(pROC)
    
    #Dummy Session with Titanic data
    titanicDF = read.csv('http://math.ucdenver.edu/RTutorial/titanic.txt',sep='\t')
    #Get the title
    titanicDF$Title = ifelse(grepl('Mr ',titanicDF$Name),'Mr',ifelse(grepl('Mrs ',titanicDF$Name),'Mrs',ifelse(grepl('Miss',titanicDF$Name),'Miss','Nothing'))) 
    #remove name and reorder
    titanicDF <- titanicDF[c('PClass', 'Age',    'Sex',   'Title', 'Survived')]

    print(str(titanicDF))
    #change to factors where needed, caret will automatically dummy code categorical variables
    titanicDF$Title = as.factor(titanicDF$Title)
    titanicDF$Survived = as.factor(titanicDF$Survived)
    #remap to Yes No with Plyr
    titanicDF$Survived = revalue(titanicDF$Survived, c("0"="No","1"="Yes"))

    #split train test
    set.seed(987)
    splitIndex <- createDataPartition(titanicDF[,"Survived"], p = .75, list = FALSE, times = 1)
    trainDF <- titanicDF[ splitIndex,]
    testDF  <- titanicDF[-splitIndex,]

    #set traincontrol - choose sparsest model within oneSE
    control <- trainControl(method="repeatedcv", number=10, repeats=10, classProbs=TRUE, summaryFunction=twoClassSummary, selectionFunction ="oneSE")

    #optimise over alpha 0.5 but range of lambda values
    tuneGrid = expand.grid(alpha=0.5, lambda = 10 ^ seq(-0.3, -5, length = 100))

    #preprocess will remove zv and nzv variables, impute missing data in age column (no other missing data in dataset) and centre/scale training data
    set.seed(987)
    titanicMod = train(Survived ~ ., data=trainDF, method="glmnet", metric="ROC", tuneGrid = tuneGrid, preProc = c("nzv","zv","center", "scale","knnImpute"), trControl=control, na.action = na.pass)
    #test the model with predict
    titanicResult = predict(titanicMod, testDF, type = "prob", na.action = na.pass)
    #get the roc auc to show model works (not required for this example)
    titanicROC = roc(predictor = titanicResult$Yes, response = testDF$Survived=="Yes")
    titanicROC$auc
    plot(titanicROC)

    #extract the best model coefficients - model has selected features
    coef(titanicMod$finalModel, titanicMod$bestTune$lambda)
    #OUTPUTS
    #8 x 1 sparse Matrix of class "dgCMatrix"
    #                   1
    #(Intercept)  -0.86345
    #PClass2nd    -0.11079
    #PClass3rd    -0.74226
    #Age          -0.15205
    #Sexmale      -0.53578
    #TitleMr      -0.48990
    #TitleMrs      0.12372
    #TitleNothing  .      

    #extract the information for standardisations of new test data
    titanicMod$preProcess$mean
    #OUTPUTS
    #PClass2nd    PClass3rd          Age      Sexmale      TitleMr     TitleMrs TitleNothing 
    #0.20994      0.54361     30.25872      0.65822      0.57911      0.14706      0.10142 

    titanicMod$preProcess$std
    #OUTPUTS
    #PClass2nd    PClass3rd          Age      Sexmale      TitleMr     TitleMrs TitleNothing 
    #0.40747      0.49835     14.37039      0.47455      0.49395      0.35434      0.30204 

    #compare predict to manually calculating the prob by just predicting first row of test data
    predict(titanicMod, testDF[1,], type = "prob", na.action = na.pass)
    #This gives a probability of survival as 0.78292

    #function to input untransformed predictor variables (those that were featured selected) for standardisation, calculation of logit and then output probability
    probSurvival = function(x_PClass2nd,x_PClass3rd,x_Age,x_Sexmale,x_TitleMr,x_TitleMrs){
  
      #Define the beta coefficients from above
      B_intercept = -0.86345
      B_PClass2nd = -0.11079
      B_PClass3rd = -0.74226
      B_Age = -0.15205
      B_Sexmale = -0.53578
      B_TitleMr =  -0.48990
      B_TitleMrs = 0.12372
  
      #standardise the predictor variables from above information for standardisation
      z_PClass2nd = (x_PClass2nd-0.20994)/0.40747
      z_PClass3rd = (x_PClass3rd-0.54361)/0.49835
      z_Age = (x_Age-30.25872)/14.37039
      z_Sexmale = (x_Sexmale-0.65822)/0.47455
      z_TitleMr = (x_TitleMr-0.57911)/0.49395 
      z_TitleMrs = (x_TitleMrs-0.14706)/0.35434
  
      #calculate logit survival
      logitSurvival = B_intercept + (B_PClass2nd * z_PClass2nd) + (B_PClass3rd * z_PClass3rd) + (B_Age * z_Age) + (B_Sexmale * z_Sexmale) + (B_TitleMr * z_TitleMr) + (B_TitleMrs * z_TitleMrs)
      #convert logit to probability
      probSurvival = 1/(1+exp(-logitSurvival))
  
      return(probSurvival)
    }

    #Testing the first row of testDF manually should give same probability of survival
    testDF[1,]
    #OUTPUTS
    #  PClass Age    Sex Title Survived
    #1    1st  29 female  Miss      Yes
    
    #Enter the values manually 
    probSurvival(0,0,29,0,0,0)
    #Gives a probability of 0.50416 which is not the same as predict!

**Comment by samleighton87** [comment]
Apologies - this code works as expected and gives the correct probability 0.7829236
